File: Tree/AVL.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int key;
    struct Node *left;
    struct Node *right;
    int height;
};

int getheight(struct Node *n)
{
    if (n == NULL)
        return 0;
    return n->height;
}

struct Node *createNode(int key)
{
    struct Node *node = (struct Node *)malloc(sizeof(struct Node));
    node->key = key;
    node->left = NULL;
    node->right = NULL;
    node->height = 1; // New node is initially at height 1
    return node;
}

int max(int a, int b)
{
    return a > b ? a : b;
}

int getBalanceFactor(struct Node *n)
{
    if (n == NULL)
        return 0;

    return getheight(n->left) - getheight(n->right);
}

struct Node *RightRotate(struct Node *y)
{
    struct Node *x = y->left;
    struct Node *T2 = x->right;

    // Perform rotation
    x->right = y;
    y->left = T2;

    // Update heights
    y->height = max(getheight(y->right), getheight(y->left)) + 1;
    x->height = max(getheight(x->right), getheight(x->left)) + 1;

    return x;
}

struct Node *LeftRotate(struct Node *x)
{
    struct Node *y = x->right;
    struct Node *T2 = y->left;

    // Perform rotation
    x->right = T2;
    y->left = x;

    // Update heights
    y->height = max(getheight(y->right), getheight(y->left)) + 1;
    x->height = max(getheight(x->right), getheight(x->left)) + 1;

    return y;
}

struct Node *insert(struct Node *node, int key)
{
    
    if (node == NULL)
        return createNode(key);

    if (key < node->key)
        node->left = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);
    else
        return node;

    node->height = 1 + max(getheight(node->left), getheight(node->right));
    int bf = getBalanceFactor(node);

    // Left Left case
    if (bf > 1 && key < node->left->key)
    {
        return RightRotate(node);
    }

    // Right Right case
    if (bf < -1 && key > node->right->key)
    {
        return LeftRotate(node);
    }

    // Left Right case
    if (bf > 1 && key > node->left->key)
    {
        node->left = LeftRotate(node->left);
        return RightRotate(node);
    }

    // Right Left case
    if (bf < -1 && key < node->right->key)
    {
        node->right = RightRotate(node->right);
        return LeftRotate(node);
    }

    return node;
}
void preOrder(struct Node *root)
{
    if (root != NULL)
    {
        printf("%d ", root->key);
        preOrder(root->left);
        preOrder(root->right);
    }
}

int main()
{
    struct Node *root = NULL;
    root = insert(root, 1);
    root = insert(root, 2);
    root = insert(root, 4);
    root = insert(root, 5);
    root = insert(root, 6);
    root = insert(root, 3);

    preOrder(root);
    printf("\n");

    return 0;
}

File: Tree/AVL.py
class Node:
    def __init__(self, key): 
        self.key = key
        self.left = None
        self.right = None
        self.height = 1

class AVLTree:
    def get_height(self, node):
        if not node:
            return 0
        return node.height

    def get_balance(self, node):
        if not node:
            return 0
        return self.get_height(node.left) - self.get_height(node.right)

    # Right Rotate
    def right_rotate(self, z):  
        y = z.left
        T3 = y.right

        y.right = z
        z.left = T3

        z.height = 1 + max(self.get_height(z.left), self.get_height(z.right))
        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))

        return y

    # Left Rotate
    def left_rotate(self, z):  
        y = z.right
        T2 = y.left

        y.left = z
        z.right = T2

        z.height = 1 + max(self.get_height(z.left), self.get_height(z.right))
        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))

        return y

    def insert(self, root, key):
        if not root:
            return Node(key)
        elif key < root.key:
            root.left = self.insert(root.left, key)
        elif key > root.key:
            root.right = self.insert(root.right, key)
        else:
            return root

        root.height = 1 + max(self.get_height(root.left), self.get_height(root.right))
        balance = self.get_balance(root)

       
        if balance > 1 and key < root.left.key:
            return self.right_rotate(root)

        if balance < -1 and key > root.right.key:
            return self.left_rotate(root)

        if balance > 1 and key > root.left.key:
            root.left = self.left_rotate(root.left)
            return self.right_rotate(root)

        if balance < -1 and key < root.right.key:
            root.right = self.right_rotate(root.right)
            return self.left_rotate(root)

        return root
# DLR
    def preorder(self, root):
        if not root:
            return
        print(root.key, end=' ')
        self.preorder(root.left)
        self.preorder(root.right)

# LRD
    def postorder(self, root):
        if not root:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.key, end=' ')
        
 # LDR
    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        print(root.key, end=' ')  
        self.inorder(root.right)
       


avl = AVLTree()
root = None

values = [10, 20, 30, 40, 50, 25, 35, 15]
for val in values:
    root = avl.insert(root, val)

print("Preorder Traversal of AVL Tree:\n")
avl.preorder(root)
print("\n")

print("inorder Traversal of AVL Tree:\n")
avl.inorder(root)
print("\n")

print("Postorder Traversal of AVL Tree:\n")
avl.postorder(root)
print("\n")
